### ISSUE TITLE
Composer would not downgrade a dependency on require

### ISSUE THREAD

**Issue by Kolyunya** [composer/composer#7329]
My `composer.json`:

```json
{
    "name": "composer/test",
    "repositories": [
        {
            "type": "composer",
            "url": "https://asset-packagist.org"
        }
    ]
}
```

Output of `composer diagnose`:

```
Checking platform settings: OK
Checking git settings: OK
Checking http connectivity to packagist: OK
Checking https connectivity to packagist: OK
Checking github.com rate limit: OK
Checking disk free space: OK
Checking pubkeys: 
Tags Public Key Fingerprint: 57815BA2 7E54DC31 7ECC7CC5 573090D0  87719BA6 8F3BB723 4E5D42D0 84A14642
Dev Public Key Fingerprint: 4AC45767 E5EC2265 2F0C1167 CBBB8A2B  0C708369 153E328C AD90147D AFE50952
OK
Checking composer version: OK
Composer version: 1.6.5
PHP version: 7.2.1
PHP binary path: /usr/bin/php
```

When I run this commands:

```
composer require codeception/codeception
composer require --update-with-all-dependencies kidzen/yii2-dynamicform
```

I get the following output:

```
Using version ^2.4 for kidzen/yii2-dynamicform
./composer.json has been updated
Loading composer repositories with package information
Updating dependencies (including require-dev)
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - Installation request for kidzen/yii2-dynamicform ^2.4 -> satisfiable by kidzen/yii2-dynamicform[v2.4].
    - Conclusion: remove symfony/css-selector v4.0.9
    - Conclusion: don't install symfony/css-selector v4.0.9
    - kidzen/yii2-dynamicform v2.4 requires symfony/css-selector ~2.8|~3.0 -> satisfiable by symfony/css-selector[v2.8.0, v2.8.1, v2.8.10, v2.8.11, v2.8.12, v2.8.13, v2.8.14, v2.8.15, v2.8.16, v2.8.17, v2.8.18, v2.8.19, v2.8.2, v2.8.20, v2.8.21, v2.8.22, v2.8.23, v2.8.24, v2.8.25, v2.8.26, v2.8.27, v2.8.28, v2.8.29, v2.8.3, v2.8.30, v2.8.31, v2.8.32, v2.8.33, v2.8.34, v2.8.35, v2.8.36, v2.8.37, v2.8.38, v2.8.39, v2.8.4, v2.8.5, v2.8.6, v2.8.7, v2.8.8, v2.8.9, v3.0.0, v3.0.1, v3.0.2, v3.0.3, v3.0.4, v3.0.5, v3.0.6, v3.0.7, v3.0.8, v3.0.9, v3.1.0, v3.1.1, v3.1.10, v3.1.2, v3.1.3, v3.1.4, v3.1.5, v3.1.6, v3.1.7, v3.1.8, v3.1.9, v3.2.0, v3.2.1, v3.2.10, v3.2.11, v3.2.12, v3.2.13, v3.2.14, v3.2.2, v3.2.3, v3.2.4, v3.2.5, v3.2.6, v3.2.7, v3.2.8, v3.2.9, v3.3.0, v3.3.1, v3.3.10, v3.3.11, v3.3.12, v3.3.13, v3.3.14, v3.3.15, v3.3.16, v3.3.2, v3.3.3, v3.3.4, v3.3.5, v3.3.6, v3.3.7, v3.3.8, v3.3.9, v3.4.0, v3.4.1, v3.4.2, v3.4.3, v3.4.4, v3.4.5, v3.4.6, v3.4.7, v3.4.8, v3.4.9].
    - Can only install one of: symfony/css-selector[v2.8.0, v4.0.9].
    - Can only install one of: symfony/css-selector[v2.8.1, v4.0.9].
    - Can only install one of: symfony/css-selector[v2.8.10, v4.0.9].
    - Can only install one of: symfony/css-selector[v2.8.11, v4.0.9].
    - Can only install one of: symfony/css-selector[v2.8.12, v4.0.9].
    - Can only install one of: symfony/css-selector[v2.8.13, v4.0.9].
    - Can only install one of: symfony/css-selector[v2.8.14, v4.0.9].
    - Can only install one of: symfony/css-selector[v2.8.15, v4.0.9].
    - Can only install one of: symfony/css-selector[v2.8.16, v4.0.9].
    - Can only install one of: symfony/css-selector[v2.8.17, v4.0.9].
    - Can only install one of: symfony/css-selector[v2.8.18, v4.0.9].
    - Can only install one of: symfony/css-selector[v2.8.19, v4.0.9].
    - Can only install one of: symfony/css-selector[v2.8.2, v4.0.9].
    - Can only install one of: symfony/css-selector[v2.8.20, v4.0.9].
    - Can only install one of: symfony/css-selector[v2.8.21, v4.0.9].
    - Can only install one of: symfony/css-selector[v2.8.22, v4.0.9].
    - Can only install one of: symfony/css-selector[v2.8.23, v4.0.9].
    - Can only install one of: symfony/css-selector[v2.8.24, v4.0.9].
    - Can only install one of: symfony/css-selector[v2.8.25, v4.0.9].
    - Can only install one of: symfony/css-selector[v2.8.26, v4.0.9].
    - Can only install one of: symfony/css-selector[v2.8.27, v4.0.9].
    - Can only install one of: symfony/css-selector[v2.8.28, v4.0.9].
    - Can only install one of: symfony/css-selector[v2.8.29, v4.0.9].
    - Can only install one of: symfony/css-selector[v2.8.3, v4.0.9].
    - Can only install one of: symfony/css-selector[v2.8.30, v4.0.9].
    - Can only install one of: symfony/css-selector[v2.8.31, v4.0.9].
    - Can only install one of: symfony/css-selector[v2.8.32, v4.0.9].
    - Can only install one of: symfony/css-selector[v2.8.33, v4.0.9].
    - Can only install one of: symfony/css-selector[v2.8.34, v4.0.9].
    - Can only install one of: symfony/css-selector[v2.8.35, v4.0.9].
    - Can only install one of: symfony/css-selector[v2.8.36, v4.0.9].
    - Can only install one of: symfony/css-selector[v2.8.37, v4.0.9].
    - Can only install one of: symfony/css-selector[v2.8.38, v4.0.9].
    - Can only install one of: symfony/css-selector[v2.8.39, v4.0.9].
    - Can only install one of: symfony/css-selector[v2.8.4, v4.0.9].
    - Can only install one of: symfony/css-selector[v2.8.5, v4.0.9].
    - Can only install one of: symfony/css-selector[v2.8.6, v4.0.9].
    - Can only install one of: symfony/css-selector[v2.8.7, v4.0.9].
    - Can only install one of: symfony/css-selector[v2.8.8, v4.0.9].
    - Can only install one of: symfony/css-selector[v2.8.9, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.0.0, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.0.1, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.0.2, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.0.3, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.0.4, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.0.5, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.0.6, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.0.7, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.0.8, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.0.9, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.1.0, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.1.1, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.1.10, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.1.2, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.1.3, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.1.4, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.1.5, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.1.6, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.1.7, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.1.8, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.1.9, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.2.0, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.2.1, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.2.10, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.2.11, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.2.12, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.2.13, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.2.14, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.2.2, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.2.3, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.2.4, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.2.5, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.2.6, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.2.7, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.2.8, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.2.9, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.3.0, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.3.1, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.3.10, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.3.11, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.3.12, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.3.13, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.3.14, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.3.15, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.3.16, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.3.2, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.3.3, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.3.4, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.3.5, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.3.6, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.3.7, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.3.8, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.3.9, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.4.0, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.4.1, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.4.2, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.4.3, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.4.4, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.4.5, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.4.6, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.4.7, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.4.8, v4.0.9].
    - Can only install one of: symfony/css-selector[v3.4.9, v4.0.9].
    - Installation request for symfony/css-selector (locked at v4.0.9) -> satisfiable by symfony/css-selector[v4.0.9].


Installation failed, reverting ./composer.json to its original content.
```

And I expected this to happen: `symfony/css-selector` to be downgraded to `v3.4.9` in order to satisfy both dependencies.

Description: both my dependencies require `symfony/css-selector`. The first installed requires `>=2.7 <5.0`. The second one requires `~2.8|~3.0`. With the `--update-with-all-dependencies` flag provided I expected the common dependency to be downgraded.

**Comment by dciancu** [comment]
Same issue here:
```shell
# Will show below error
➜ composer create-project --prefer-dist laravel/laravel laravel-test && cd laravel-test && composer require --dev codeception/aspect-mock
# Will work
➜ mv composer.lock composer.lock.bak && mv vendor vendor.bak && composer require --dev codeception/aspect-mock 
```

Error
```
Using version ^3.0 for codeception/aspect-mock
./composer.json has been updated
Loading composer repositories with package information
Updating dependencies (including require-dev)
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - Conclusion: don't install codeception/aspect-mock 3.0.2
    - Conclusion: don't install codeception/aspect-mock 3.0.1
    - Conclusion: remove nikic/php-parser v4.1.0
    - Installation request for codeception/aspect-mock ^3.0 -> satisfiable by codeception/aspect-mock[3.0.0, 3.0.1, 3.0.2].
    - Conclusion: don't install nikic/php-parser v4.1.0
    - codeception/aspect-mock 3.0.0 requires goaop/framework ^2.2.0 -> satisfiable by goaop/framework[2.2.0, 2.x-dev].
    - goaop/framework 2.2.0 requires goaop/parser-reflection ~1.4 -> satisfiable by goaop/parser-reflection[1.4.0, 1.4.1, 1.x-dev].
    - goaop/framework 2.x-dev requires goaop/parser-reflection ~1.4 -> satisfiable by goaop/parser-reflection[1.4.0, 1.4.1, 1.x-dev].
    - goaop/parser-reflection 1.4.0 requires nikic/php-parser ^1.2|^2.0|^3.0 -> satisfiable by nikic/php-parser[1.x-dev, 2.x-dev, 3.x-dev, v1.2.0, v1.2.1, v1.2.2, v1.3.0, v1.4.0, v1.4.1, v2.0.0, v2.0.0alpha1, v2.0.0beta1, v2.0.1, v2.1.0, v2.1.1, v3.0.0, v3.0.0alpha1, v3.0.0beta1, v3.0.0beta2, v3.0.1, v3.0.2, v3.0.3, v3.0.4, v3.0.5, v3.0.6, v3.1.0, v3.1.1, v3.1.2, v3.1.3, v3.1.4, v3.1.5].
    - goaop/parser-reflection 1.4.1 requires nikic/php-parser ^1.2|^2.0|^3.0 -> satisfiable by nikic/php-parser[1.x-dev, 2.x-dev, 3.x-dev, v1.2.0, v1.2.1, v1.2.2, v1.3.0, v1.4.0, v1.4.1, v2.0.0, v2.0.0alpha1, v2.0.0beta1, v2.0.1, v2.1.0, v2.1.1, v3.0.0, v3.0.0alpha1, v3.0.0beta1, v3.0.0beta2, v3.0.1, v3.0.2, v3.0.3, v3.0.4, v3.0.5, v3.0.6, v3.1.0, v3.1.1, v3.1.2, v3.1.3, v3.1.4, v3.1.5].
    - goaop/parser-reflection 1.x-dev requires nikic/php-parser ^1.2|^2.0|^3.0 -> satisfiable by nikic/php-parser[1.x-dev, 2.x-dev, 3.x-dev, v1.2.0, v1.2.1, v1.2.2, v1.3.0, v1.4.0, v1.4.1, v2.0.0, v2.0.0alpha1, v2.0.0beta1, v2.0.1, v2.1.0, v2.1.1, v3.0.0, v3.0.0alpha1, v3.0.0beta1, v3.0.0beta2, v3.0.1, v3.0.2, v3.0.3, v3.0.4, v3.0.5, v3.0.6, v3.1.0, v3.1.1, v3.1.2, v3.1.3, v3.1.4, v3.1.5].
    - Can only install one of: nikic/php-parser[1.x-dev, v4.1.0].
    - Can only install one of: nikic/php-parser[2.x-dev, v4.1.0].
    - Can only install one of: nikic/php-parser[3.x-dev, v4.1.0].
    - Can only install one of: nikic/php-parser[v1.3.0, v4.1.0].
    - Can only install one of: nikic/php-parser[v1.4.0, v4.1.0].
    - Can only install one of: nikic/php-parser[v1.4.1, v4.1.0].
    - Can only install one of: nikic/php-parser[v2.0.0, v4.1.0].
    - Can only install one of: nikic/php-parser[v2.0.0alpha1, v4.1.0].
    - Can only install one of: nikic/php-parser[v2.0.0beta1, v4.1.0].
    - Can only install one of: nikic/php-parser[v2.0.1, v4.1.0].
    - Can only install one of: nikic/php-parser[v2.1.0, v4.1.0].
    - Can only install one of: nikic/php-parser[v2.1.1, v4.1.0].
    - Can only install one of: nikic/php-parser[v3.0.0, v4.1.0].
    - Can only install one of: nikic/php-parser[v3.0.0alpha1, v4.1.0].
    - Can only install one of: nikic/php-parser[v3.0.0beta1, v4.1.0].
    - Can only install one of: nikic/php-parser[v3.0.0beta2, v4.1.0].
    - Can only install one of: nikic/php-parser[v3.0.1, v4.1.0].
    - Can only install one of: nikic/php-parser[v3.0.2, v4.1.0].
    - Can only install one of: nikic/php-parser[v3.0.3, v4.1.0].
    - Can only install one of: nikic/php-parser[v3.0.4, v4.1.0].
    - Can only install one of: nikic/php-parser[v3.0.5, v4.1.0].
    - Can only install one of: nikic/php-parser[v3.0.6, v4.1.0].
    - Can only install one of: nikic/php-parser[v3.1.0, v4.1.0].
    - Can only install one of: nikic/php-parser[v3.1.1, v4.1.0].
    - Can only install one of: nikic/php-parser[v3.1.2, v4.1.0].
    - Can only install one of: nikic/php-parser[v3.1.3, v4.1.0].
    - Can only install one of: nikic/php-parser[v3.1.4, v4.1.0].
    - Can only install one of: nikic/php-parser[v3.1.5, v4.1.0].
    - Can only install one of: nikic/php-parser[v1.2.0, v4.1.0].
    - Can only install one of: nikic/php-parser[v1.2.1, v4.1.0].
    - Can only install one of: nikic/php-parser[v1.2.2, v4.1.0].
    - Installation request for nikic/php-parser (locked at v4.1.0) -> satisfiable by nikic/php-parser[v4.1.0].


Installation failed, reverting ./composer.json to its original content.
```

**Comment by BackEndTea** [comment]
For those who run into this ticket while searching for a solution, you can add the specific version you need to solve this to the require.

(In my case i wanted to install drupal/console, but i had some conflicting symfony packages)

First i ran the following:
```bash
 composer require --dev drupal/console symfony/filesystem:^3.4 symfony/finder:^3.4 symfony/dom-crawler:^3.4
```

Which installed all the packages, at their correct version. Replace these with the package you want to intall/ the version constraints you need.

Post install my `composer.json` looked like this:
```json
"require-dev": {
    "drupal/console": "^1.9",
    "symfony/filesystem": "^3.4",
    "symfony/finder": "^3.4",
    "symfony/dom-crawler": "^3.4"
},
```

Now remove the manually added constraints:
```json
"require-dev": {
    "drupal/console": "^1.9"
},
```

and finally run `composer update --lock`, which makes sure your lockfile is once again updated, but without updating new dependencies.

**Comment by naderman** [comment]
@Seldaek I think this may be fixed with your replace listing change for https://github.com/composer/composer/issues/8355 ?

**Comment by Seldaek** [comment]
This is fixed now by #8717
